src/dmd/mars.d

```diff
@@ -1303,6 +1303,20 @@ void addDefaultVersionIdentifiers()
         VersionCondition.addPredefinedGlobalIdent("FreeBSD");
         VersionCondition.addPredefinedGlobalIdent("ELFv1");
         global.params.isFreeBSD = true;
+
+        import core.sys.posix.sys.utsname : uname, utsname;
+        utsname utsn;
+        if (uname(&utsn) == 0)
+        {
+            foreach (i, c; utsn.release)
+            {
+                if (c >= '0' && c <= '9')
+                    continue;
+                if (c == '.')
+                    VersionCondition.addPredefinedGlobalIdent("FreeBSD" ~ utsn.release[0 .. i]);
```

Comment:
Anything versioned under `version(FreeBSD12)` would then not be versioned in with FreeBSD 13, but we could either have dropped the FreeBSD 11 bindings by that point so that we wouldn't need `version(FreeBSD12)`, or it would be simple enough to duplicate the 12 bindings under `version(FreeBSD13)`. If we're not going to use some sort of conditional, then anything that's versioned for a specific version is inherently going to break with new versions unless it's under `version(FreeBSD)`, and it's just the older bindings using a specific version (e.g. `version(FreeBSD11)`).

Regardless, I think that we're talking about fewer than half a dozen version statements with regards to the inode64 changes. Maybe there are a few other places where something changed, but _most_ of the bindings stay the same from version to version. So, this really isn't going to be much of a maintenance burden, and since FreeBSD is my main OS, I'm definitely going to be making sure that this stuff continues to work even beyond whatever the auto-tester is currently testing. The bigger problem is that without being able to version the bindings based on the OS version at all, we can't have D working on both FreeBSD 11 and 12 at the same time even temporarily, making transitioning a pain for both the release team and for users.

Reply:
How about instead of versioning FreeBSD itself, just adding a version for that specific feature? Then you wouldn't need to add a `FreeBSD13` symbol.

```d
version (FreeBSD_32BitINodeScheme) /* or whatever name it should have */
{
    // ...
}
else
{
    // ...
}
```

Reply:
Fundamentally, the bindings go with a specific version of FreeBSD, so I really don't think that it makes sense to try and version them based on feature changes. _Why_ the changes were made is pretty much irrelevant really. It's the simple fact that the bindings changed due to a change in the headers - which comes with each version of the OS - that matters.

And I specifically wrote this PR in a way that made it so that we just permanently get the appropriate version identifier for the current version without having make any changes to dmd, whereas if we came up with version identifiers for particular features, we'd have to keep mucking with the version identifiers in dmd whenever we needed to version something due to an OS change. And while hopefully the need to version bindings is rare, I'd much rather set up the version identifiers in a way that minimizes the amount of maintenance required rather than having to add a new one every time the bindings need to be updated in a backwards incompatible way.

Reply:
Let's just do what FreeBSD finds appropriate ;), `__getosreldate() >= INO64_FIRST` ([INO64_FIRST](https://github.com/freebsd/freebsd/blob/25c1ae6892ac8a22e7f8a9cf2e6b6ccc4d0841ff/lib/libc/sys/compat-ino64.h#L43) == 1200031).
We could expose `__getosreldate()` as `__FreeBSD_version` lexer macro which would match their C `osreldate.h` header. Or add a `core.sys.freebsd.osreleasdate` module that has a `__FreeBSD_version` enum (still needed a lexer macro to get the value from the compiler).
Also see [share/examples/FreeBSD_version](https://github.com/freebsd/freebsd/blob/386ddae58459341ec567604707805814a2128a57/share/examples/FreeBSD_version/FreeBSD_version.c).

Making releases available as versions identifiers seem messy due to the fact that they are not comparable.

Reply:
If we want to match the C headers, then we pretty much need to version based on the major release, since aside from what happens to FreeBSD CURRENT during development, that's exactly what you get with the C headers. Also, it's not necessarily the case that the bindings need to change only for the ino64 changes. Those are the changes that broke dmd on FreeBSD 12, but other bindings may have changed and still need to be versioned (thus far, I've just fixed the ones necessary to get dmd working), and others could change for FreeBSD 13. By versioning based on the OS version, we match what C/C++ is doing.

We could certainly choose to expose the result of `__getosreldate()` at compile time and use static ifs to version the bindings rather than version identifiers, but aside from issues with trying to support FreeBSD CURRENT as it changed (which we probably don't want to worry about if we can avoid it), the version identifiers work just fine (we just have to duplicate the occasional binding just like we already do with different OS bindings), and Walter is usually very opposed to versioning code based on any kind of conditional logic, so I'd be surprised if he were very happy with having something like

    static if(___FreeBSD_version > INO_64)

in druntime. It also actually would make it harder to version the bindings for any future OS changes. In the case of the ino64 changes, they were a big enough deal, that we know precisely which value of `__getosreldate()` to compare against, but for other API changes, we'd basically have to track FreeBSD development to know exactly when any APIs changed, whereas if we're using stuff like `version(FreeBSD12)`, then we can just worry about the actual releases and maybe go to the effort of updating the bindings for FreeBSD 13 every once in a while (if necessary) to ensure that distros like TrueOS and GhostBSD continue to work, and we wouldn't have to worry about when exactly a particular change was introduced.

If the consensus is that we want to expose the result of `__getosreldate() through an enum that can be used to version the bindings, then I'm okay with that, but aside from trying to support changes to FreeBSD CURRENT, it doesn't seem like a better solution to me. I also don't know enough about dmd to implement it, so either someone else would need to do it, or I would have to spend the time necessary to figure it out. I can't simply update this PR to use that approach, since I don't currently have the knowledge to do so.

Reply:
Another reason to use the internal version instead of the publicly released version number is that it makes bisecting of the entire system possible. E.g. let's say a large and complicated D program works on FreeBSD 11 but not on 12 - if you look at only the major version, you can't bisect the system to find which change in FreeBSD broke your D program.

> It also actually would make it harder to version the bindings for any future OS changes. In the case of the ino64 changes, they were a big enough deal, that we know precisely which value of `__getosreldate()` to compare against, but for other API changes, we'd basically have to track FreeBSD development to know exactly when any APIs changed,

How likely is this scenario to occur, exactly? Can you give some examples where we can't use the same approach for FreeBSD changes that affect D?